Patch-Source: https://github.com/nix-rust/nix/pull/1900
Paths were adapted for the vendor/ directory.
---
From 3c25dbcb22e777c0d40e892d3f4229bb2f36608b Mon Sep 17 00:00:00 2001
From: Wesley Wiser <wesleywiser@microsoft.com>
Date: Thu, 1 Dec 2022 19:07:11 -0500
Subject: [PATCH] Update use of libc::timespec to prepare for future libc
 version

This is a backport of 006fc6f7975b3a6b64329847b780622aab392109. The
original commit message follows:

In a future release of the `libc` crate, `libc::timespec` will contain
private padding fields on `*-linux-musl` targets and so the struct will
no longer be able to be created using the literal initialization syntax.

Update places where `libc::timespec` is created to first zero initialize
the value and then update the `tv_sec` and `tv_nsec` fields manually.
Many of these places are in `const fn`s so a helper function
`zero_init_timespec()` is introduced to help with this as
`std::mem::MaybeUninit::zeroed()` is not a `const` function.

Some matches on `libc::timespec` are also updated to include a trailing
`..` pattern which works when `libc::timespec` has additional, private
fields as well as when it does not (like for
`x86_64-unknown-linux-gnu`).
---
 src/sys/time.rs    | 31 +++++++++++++++++++++----------
 src/sys/timerfd.rs | 18 +++++-------------
 2 files changed, 26 insertions(+), 23 deletions(-)

diff --git a/src/sys/time.rs b/src/sys/time.rs
index 7546d1b367..5655039ce4 100644
--- a/vendor/nix/src/sys/time.rs
+++ b/vendor/nix/src/sys/time.rs
@@ -5,6 +5,10 @@ use libc::{timespec, timeval};
 #[cfg_attr(target_env = "musl", allow(deprecated))] // https://github.com/rust-lang/libc/issues/1848
 pub use libc::{time_t, suseconds_t};
 
+pub(crate) const TIMESPEC_ZERO: libc::timespec = unsafe {
+    std::mem::transmute([0u8; std::mem::size_of::<libc::timespec>()])
+};
+
 pub trait TimeValLike: Sized {
     #[inline]
     fn zero() -> Self {
@@ -76,12 +80,12 @@ impl From<timespec> for TimeSpec {
 }
 
 impl From<Duration> for TimeSpec {
+    #[cfg_attr(target_env = "musl", allow(deprecated))] // https://github.com/rust-lang/libc/issues/1848
     fn from(duration: Duration) -> Self {
-        #[cfg_attr(target_env = "musl", allow(deprecated))] // https://github.com/rust-lang/libc/issues/1848
-        TimeSpec(timespec {
-            tv_sec: duration.as_secs() as time_t,
-            tv_nsec: duration.subsec_nanos() as timespec_tv_nsec_t
-        })
+        let mut ts = TIMESPEC_ZERO;
+        ts.tv_sec = duration.as_secs() as time_t;
+        ts.tv_nsec = duration.subsec_nanos() as timespec_tv_nsec_t;
+        TimeSpec(ts)
     }
 }
 
@@ -123,11 +127,15 @@ impl PartialOrd for TimeSpec {
 
 impl TimeValLike for TimeSpec {
     #[inline]
+    #[cfg_attr(target_env = "musl", allow(deprecated))]
+    // https://github.com/rust-lang/libc/issues/1848
     fn seconds(seconds: i64) -> TimeSpec {
         assert!(seconds >= TS_MIN_SECONDS && seconds <= TS_MAX_SECONDS,
                 "TimeSpec out of bounds; seconds={}", seconds);
-        #[cfg_attr(target_env = "musl", allow(deprecated))] // https://github.com/rust-lang/libc/issues/1848
-        TimeSpec(timespec {tv_sec: seconds as time_t, tv_nsec: 0 })
+        let mut ts = TIMESPEC_ZERO;
+        ts.tv_sec = seconds as time_t;
+        ts.tv_nsec = 0;
+        TimeSpec(ts)
     }
 
     #[inline]
@@ -149,13 +157,16 @@ impl TimeValLike for TimeSpec {
 
     /// Makes a new `TimeSpec` with given number of nanoseconds.
     #[inline]
+    #[cfg_attr(target_env = "musl", allow(deprecated))]
+    // https://github.com/rust-lang/libc/issues/1848
     fn nanoseconds(nanoseconds: i64) -> TimeSpec {
         let (secs, nanos) = div_mod_floor_64(nanoseconds, NANOS_PER_SEC);
         assert!(secs >= TS_MIN_SECONDS && secs <= TS_MAX_SECONDS,
                 "TimeSpec out of bounds");
-        #[cfg_attr(target_env = "musl", allow(deprecated))] // https://github.com/rust-lang/libc/issues/1848
-        TimeSpec(timespec {tv_sec: secs as time_t,
-                           tv_nsec: nanos as timespec_tv_nsec_t })
+        let mut ts = TIMESPEC_ZERO;
+        ts.tv_sec = secs as time_t;
+        ts.tv_nsec = nanos as timespec_tv_nsec_t;
+        TimeSpec(ts)
     }
 
     fn num_seconds(&self) -> i64 {
diff --git a/src/sys/timerfd.rs b/src/sys/timerfd.rs
index 44915be1f9..b6303c91f6 100644
--- a/vendor/nix/src/sys/timerfd.rs
+++ b/vendor/nix/src/sys/timerfd.rs
@@ -28,7 +28,7 @@
 //! // We wait for the timer to expire.
 //! timer.wait().unwrap();
 //! ```
-use crate::sys::time::TimeSpec;
+use crate::sys::time::{TIMESPEC_ZERO, TimeSpec};
 use crate::unistd::read;
 use crate::{errno::Errno, Result};
 use bitflags::bitflags;
@@ -89,14 +89,8 @@ struct TimerSpec(libc::itimerspec);
 impl TimerSpec {
     pub fn none() -> Self {
         Self(libc::itimerspec {
-            it_interval: libc::timespec {
-                tv_sec: 0,
-                tv_nsec: 0,
-            },
-            it_value: libc::timespec {
-                tv_sec: 0,
-                tv_nsec: 0,
-            },
+            it_interval: TIMESPEC_ZERO,
+            it_value: TIMESPEC_ZERO,
         })
     }
 }
@@ -111,10 +105,7 @@ impl From<Expiration> for TimerSpec {
     fn from(expiration: Expiration) -> TimerSpec {
         match expiration {
             Expiration::OneShot(t) => TimerSpec(libc::itimerspec {
-                it_interval: libc::timespec {
-                    tv_sec: 0,
-                    tv_nsec: 0,
-                },
+                it_interval: TIMESPEC_ZERO,
                 it_value: *t.as_ref(),
             }),
             Expiration::IntervalDelayed(start, interval) => TimerSpec(libc::itimerspec {
@@ -137,6 +128,7 @@ impl From<TimerSpec> for Expiration {
                     libc::timespec {
                         tv_sec: 0,
                         tv_nsec: 0,
+                        ..
                     },
                 it_value: ts,
             }) => Expiration::OneShot(ts.into()),
